### ISSUE TITLE
Use Cross-Entropy instead of MSE for value loss

### ISSUE THREAD

**Issue by jpdoyle** [leela-zero/leela-zero#1610]
This might be a suspect or controversial suggestion, but I think it has potential to seriously improve LZ's performance.

TL;DR the title -- if you want to interpret the value output as a "probability of winning", L2/MSE is simply *wrong*, and we should be using Cross-Entropy. Yes, MSE is how Alpha{Go,GoZero,Zero} do it. I think they've made a mistake, but are getting away with it because they have a lot of compute, a lot of genius design in other components, and a head start.

Losses provide a differentiable "wrongness" score, but to understand my suggestion we have to talk about what "wrongness" means in machine learning.

Background
==========

First, let's do some statistical inference. Suppose you have a continuous random variable `X` and you want to estimate its mean from a bunch of samples. The obvious way is to take those values, and average them. This works pretty well.

What if you have `(X,Y)` pairs instead, and want to model `Y` as a function of `X`? Well you don't have any uncertainty in the `X` values, but you might expect to have measurement error in your values of `Y` (you could have error in both terms, but this is messier and we won't worry about that). If a variable is continuous, you can usually expect the error to follow a normal (bell-curve) distribution, and the probability of a certain error value `e` is `p(E = e) = (1/sqrt(2*pi))*exp(-e^2/2)` (technically this is a probability *density* but the distinction isn't vital here). Let's assume that `Y` has a linear relationship with `X`, and we can now write `Y = F(X) + E`, where `F(x) = a*x + b` is our linear approximator, and `E` is our error term. Since `F` is a non-random function, we can rewrite that equation and get the probability of a particular `(x,y)` pair occurring:

    Y = F(X) + E
    E = Y - F(X)
    p((X,Y) = (x,y)) = p(E = (y-F(x)))

Where this gets interesting (and relevant to deep learning) is trying to find the values for `a` and `b` from a series of samples. If we pretend that they're random variables `A` and `B`, we can model their conditional probabilities given data `p((A,B) = (a,b) | X,Y)` -- read this as "the probability that `A` and `B` are `a` and `b` respectively, given the observed values of `X` and `Y`" (`|` is usually read as "given"/"given that"). If we can estimate this probability, we can pick better values of `a` and `b` to predict the data better.

So in comes Bayes' Rule, a handy little bit of elementary probability theory:

    p(A|B) = p(B|A)p(A)/p(B)

(these are generic `p`, `A` and `B`, not our specific ones).

If we can estimate `p(X,Y | A,B)` (the likelihood of data *assuming our choices of `a` and `b` are correct*), `p(A,B)` (the base, or "prior" likelihood of picking certain values for `a` and `b`), and `p(X,Y)` (the likelihood of a pair of data samples matching our data), we can get `p((A,B) = (a,b) | X,Y)`! Unfortunately, if we knew `p(X,Y)`, we'd already have our problem solved. But! We don't really need "the probability of `a` and `b`" for training, we just need a way to *improve* the probability. So we can do a little trick that makes `p(X,Y)` irrelevant by taking the logarithm of this equation, and then its partial derivatives (to get a gradient):

    p((A,B) = (a,b) | X,Y) = p(X,Y | (A,B) = (a,b))*p((A,B) = (a,b))/p(X,Y)
    ln p((A,B) = (a,b) | X,Y) = ln p(X,Y | (A,B) = (a,b)) + ln p((A,B) = (a,b)) - ln p(X,Y)
    d/da ln p((A,B) = (a,b) | X,Y) = d/da ln p(X,Y | (A,B) = (a,b)) + d/da ln p((A,B) = (a,b)) + 0
    d/db ln p((A,B) = (a,b) | X,Y) = d/db ln p(X,Y | (A,B) = (a,b)) + d/db ln p((A,B) = (a,b)) + 0

So after all this, we have a gradient relative to our single "weight" (`a`) and our "bias" (`b`), and we just have to define our `p`s and apply gradient ascent (because you want to *increase* the probability). If you negate it, the term `-ln p(X,Y | (A,B) = (a,b))` is the familiar "loss", and `-ln p((A,B) = (a,b))` is the "regularizer". This trick with Bayes' rule is at the core of deep learning.

MSE vs Cross-Entropy
====================

MSE is usually written as (in python-esque notation) `sum(((y - f(x))**2 for (x,y) in batch))`, or sometimes `sum((((y - f(x))**2)/2 for (x,y) in batch))`. Where does the `/2` come from? Well, let's plug it into our trick above:

    -ln p((X,Y) = (x,y) | F = f) = sum((((y - f(x))**2)/2 for (x,y) in batch))
    p((X,Y) = (x,y) | F = f) = exp(-sum((((y - f(x))**2)/2 for (x,y) in batch)))
    p((X,Y) = (x,y) | F = f) = product((exp(-((y - f(x))**2)/2) for (x,y) in batch))

See the `exp(-((y - f(x))**2)/2)`? That's (proportional to) the normal distribution! So MSE is saying "we expect the error to be normally distributed around zero". This is totally fine for typical, continuous functions. However, the win result of Go (at least, under the rules LZ uses) can only be `B wins` or `W wins`, and isn't continuous. That should already make you suspicious. 

Cross-Entropy loss is more popular for classification tasks, and looks like `sum(-ln(f(x)[y]) for (x,y) in batch))`. In this case, `y` is a class label, and `f(x)` is an array of probabilities for each class. If `y` is only zero or one, we can rewrite it as `sum((((1-y)*-ln(1-f(x)) + y*-ln(f(x))) for (x,y) in batch))`. What distribution does this line up with?

    -ln p((X,Y) = (x,y) | F = f) = sum((((1-y)*-ln(1-f(x)) + y*-ln(f(x))) for (x,y) in batch))
    p((X,Y) = (x,y) | F = f) = exp(-sum((((1-y)*-ln(1-f(x)) + y*-ln(f(x))) for (x,y) in batch)))
    p((X,Y) = (x,y) | F = f) = exp(sum(((-(1-y)*-ln(1-f(x)) + -y*-ln(f(x))) for (x,y) in batch)))
    p((X,Y) = (x,y) | F = f) = product((exp((1-y)*ln(1-f(x))) * exp(y*ln(f(x))) for (x,y) in batch))
    p((X,Y) = (x,y) | F = f) = product((((1-f(x))^(1-y) * f(x)^y) for (x,y) in batch))
    p((X,Y) = (x,y) | F = f) = product(((f(x) if y == 1 else (1-f(x))) for (x,y) in batch))

This is the [Bernoulli Distribution](https://en.wikipedia.org/wiki/Bernoulli_distribution) -- ie, a biased coin-flip. If you want to treat a value as a "probability of winning from this point", the bernoulli distribution is the right choice.

(To give credit where it is due, the first place I saw this interpretation of deep learning was [Alex Terenin's blog post on the subject](http://avt.im/blog/2017/07/05/bayesian-learning))

Does it matter that much?
=========================

In theory, given infinite time and a perfect model, these two loss functions should actually give the same result! This is because [the mean of an indicator variable is its probability of being 1](http://statweb.stanford.edu/~susan/courses/s116/node102.html). However, neural networks aren't perfect, and the training has to stop at some point, so the choice of loss can make a big difference in practice.

However, I don't really *know* that this will improve things! I have a hunch, and I'm hoping someone with enough compute available will be generous enough to test my hypothesis.

What other changes need to be made to evaluate this?
====================================================

None! MCTS is written assuming that the value function outputs a probability of winning, so hopefully this will just make the MCTS estimates better and LZ stronger overall.

**Comment by Eddh** [comment]
I guess this could be tested by training 2 supervised nets using the different loss functions ?

**Comment by gcp** [comment]
>I guess this could be tested by training 2 supervised nets using the different loss functions ?

Yes.

In my experience, if you can get MSE to converge at all, it's not worse. There are network designs where cross-entropy converges easily and MSE is much harder to get going. But the AZ ones don't seem to be among them.

You should just implement this (trivial change to TensorFlow) and demonstrate that it works better, the theoretical argument is well known and really not interesting at all.

**Comment by jpdoyle** [comment]
> I guess this could be tested by training 2 supervised nets using the different loss functions ?

If I understand how the supervised training data set is generated (I may not, I haven't dug into the code yet), the answer is "kinda". The value network's results should be slightly improved, but if you don't re-do the search, you will still have a computed policy distribution based on an MSE-trained value network. You may not see dramatic results until new self-play games generate searches based on the differently-trained value network.

> In my experience, if you can get MSE to converge at all, it's not worse.

Yep, this is often true for classification tasks! But most classification tasks will use argmax and just pick the highest-scored class. LZ and AZ treat the scores as probabilities, so I'd be cautious to apply that heuristic here.

> You should just implement this (trivial change to the TensorFlow code) and demonstrate that it works better

I don't personally have the computing resources to do this training and empirically justify it. That's why I'm asking here. I'd run my own fork to test it if I could.

> the theoretical argument is well known and really not interesting at all.

My friends who do Bayesian inference would disagree with "not interesting" ;)

> I'm sure the people at DeepMind are very aware that Cross-Entropy is a thing.

Well yeah, they're using it everywhere. But they're also using L2 for the value loss in all their papers without explanation or justification. It's easy to miss/dismiss something small like this when you're focusing on making the big improvements like AlphaGo Zero's policy loss.

> I'm pretty sure this has been brought up several times, too.

I found a related question in the thread for #1480. If you (and others with computing resources) aren't interested interested in experimenting with this, that's fair, but I think it's worth a shot.

**Comment by gcp** [comment]
>but I think it's worth a shot.

That's good but I can only observe that no-one (including you) has bothered testing it since it was posted months ago.

**Comment by jpdoyle** [comment]
> I don't understand why you believe this could even be an issue, or why it would be relevant for doing such a test. What shape or distribution change would you expect a different value target to cause that affects the search probabilities?

The training data is (board state, game result, generated tree-search based probabilities), no? Using cross-entropy for value training could *only* change the value function unless it improves some sort of latent-space properties in the shared network. That is, without new training data generated with the new network, the policy would still be trained to match the next-move distribution generated from the old value network. If the training does its own searching/pondering, this isn't an issue, but in a cursory glance at the training code this doesn't seem to be the case.

> That's good but I can only observe that no-one (including you) has bothered testing it since it was originally posted months ago.

I can't speak for others, but it seems a bit unfair to attack the idea simply because no one has bothered testing it. Even though it's a one-line change, running the training and evaluating it appears to be an enormous effort which demands a large amount of computing power and time -- even for those who know the code much better than I.

Do you or other folks on the project have resources I could use so I can learn how to run this training and evaluation, perhaps in the Google Colab environment? I don't have a lot of time to contribute, but if no one on the team wants to run the test I can try to take some steps toward that on my own.

**Comment by herazul** [comment]
>I can't speak for others, but it seems a bit unfair to attack the idea simply because no one has bothered testing it. 

As someone who followed the project from start (pure curiosity, didn't contribute anything), i understand why he answered that way : It's not a personnal attack against you.
There is a lot of theory-craft of dozens of ideas on the sub. He DOESN'T have the time to code, try, and test them all, it's impossible for a human alone to do it. So the only way it can work is by presentiong results and evidences that your idea work, and then, and only then, GCP can take time to work and implement it.
There is no other way.

**Comment by ThorAvaTahr** [comment]
In addition, though many ideas are theoretically sound and may in fact be good improvements, the current set up is flight-proven and strongly optimized. 
I have personally experienced, that it is very hard to compete with that setup. Any change would almost certainly decrease strength, before it can Improve anything.
Therefore i strongly support the, sometimes harsh, dogma by GCP that only theory that comes with extensive testing and most of all hard data is elligible to be considered to be included. I have had even seemingly small changes been refused, just because consequences are unforeseen and insufficiently tested.

**Comment by gcp** [comment]
>That is, without new training data generated with the new network, the policy would still be trained to match the next-move distribution generated from the old value network

That's fine. If the method is an improvement, the improved value targets would make it stronger. They can shift the policy targets in the next iterations. If the value network is not an improvement, it won't help policy iterations. So you should still be able to detect improvement from the value net only.

>it seems a bit unfair to attack the idea

I'm not attacking the idea. I'm pointing out that unless someone tests it, no amount of writing or theorizing here is going to change anything. Even if you are 100% correct.

>Do you or other folks on the project have resources I could use so I can learn how to run this training and evaluation, perhaps in the Google Colab environment? I don't have a lot of time to contribute, but if no one on the team wants to run the test I can try to take some steps toward that on my own.

You realistically need an NVIDIA GPU (in theory using the CPU works, but in practice the slowdown is too big) and a network connection that can download a few gigabyte of data, and that's about it. It doesn't need to be a top of the line one, you could verify the idea works with a smaller network first, or with a bit more patience.

If you have that, fetch the github repo, set up a python enviroment in the /training/tf dir and install the requirements there (pip install -r requirements.txt which will get you TensorFlow). Get about 500k worth of games from #167 and follow the instructions on the readme.

**Comment by Marcin1960** [comment]
@jpdoyle "up until after that point, but I'll see what I can do! "

By  all means, do it. One can wait :)

**Comment by godmoves** [comment]
I think something similar is mentioned in [this paper](https://arxiv.org/pdf/1705.08439.pdf), they invent an algorithm to play the game Hex, which is similar to the alphago zero one, but use cross entropy rather than MSE loss for the value part. So I think cross entropy may also work for this kind of problems, although they don't test the differences in performance.

Another interesting thing is their result shows that keeping all data for training is slightly better than using a fix moving window.

**Comment by tterava** [comment]
I'd like some help testing this out. I've downloaded roughly 500k latest self play games and I'm trying to modify the training to use cross-entropy as suggested. However I'm not sure how to modify the training. I tried doing this:

```
# Loss on value head
# self.mse_loss = \
#     tf.reduce_mean(tf.squared_difference(self.z_, self.z_conv))

value_cross_entropy = \
    tf.nn.softmax_cross_entropy_with_logits(labels=self.z_,
                                            logits=self.z_conv)
self.mse_loss = tf.reduce_mean(value_cross_entropy)
    
```
Basically i just copied what the policy head was doing as it's already using cross entropy. However the value head reports 0 loss during training so there's something wrong here. The variable names are a bit confusing but i kept the mse_loss so I wouldn't have to change it to every other place. I did remove the scaling from the reporting though. Learning rate was set to 0.005 as instructed in the comments.

I would also like some input on what would be the best way to test if this improves accuracy. I was thinking of taking one of the early 192x15 nets (ba748d40) and training it for X amount of steps with the latest selfplay games and then see if it converges quicker. Does this make sense? How many steps should I train for?

**Comment by Ttl** [comment]
As the function name hints, `softmax_cross_entropy_with_logits` calculates cross entropy of softmax logits. Since you have only one class taking softmax gives 1 for every input.

Take a look at `tf.nn.sigmoid_cross_entropy_with_logits` instead. Note that you need to scale labels to [0, 1] since they represent a probability distribution. `logits` must be taken before the `tanh` of the value head since loss calculation takes sigmoid of `logits`. You also need to scale the logits so that it works correctly in LZ. `sigmoid(2x) = (1 + tanh(x))/2`.

I suggest instead training two small networks from scratch with different value head loss while keeping the other parameters constant. Then you can match the two networks together to see which one is better.

**Comment by tterava** [comment]
It's been quite some time when I last worked with TF and i misunderstood that labels already had 2 classes. So am I understanding correctly that first I have to rescale and shape the label so that `1  ->  [1, 0]` and `-1  ->  [0, 1]`? Then I can do the same with the value head and get loss with `softmax_cross_entropy_with_logits` ?

**Comment by tterava** [comment]
Ok it's a bit of a hack but here's what I did:
`v_labels = tf.concat([(self.z_ + 1.0) / 2.0, (self.z_ - 1.0) / -2.0], 1)`
`v_logits = tf.concat([(self.z_conv + 1.0) / 2.0, (self.z_conv - 1.0) / -2.0], 1)`

This turns `[ [1], [-1], [1] ... ]` into `[ [1, 0], [0, 1], [1, 0] ... ]`
Or [ [0.8], [-0.5] ... ] into [ [0.9, 0.1], [0.25, 0.75] ... ]

If i understood correctly, self.z_conv also gives win percentage in range [-1, 1] as well so this should work for `softmax_cross_entropy_with_logits`. I hope this makes sense.

I switched to a smaller 128x10 net and I'll try training one from scratch. After 10k steps the loss has dropped from initial 0.7 (unscaled) to 0.58 so it seems to be working.

**Comment by tterava** [comment]
I trained 2 networks of size 128x10 for 200 000 steps. I'll attach all the logs and some networks along the way.

I'm not an expert analysing the data but I did some quick testing (-r 5 -v 300) with some of the networks along the way and it seems the early cross entropy networks are stronger than their MSE counter parts. This would lead me to believe that using cross entropy makes the value head converge faster. However it doesn't seem too different than just using a higher learning rate with MSE. The final 200k network had an SPRT fail with about 35% winrate.

I don't know how to further test this or what to really look for. If these results aren't conclusive, I would appreciate some advice on what to do next.

NOTE: Logs of cross entropy training mistakenly still label it as mse. I didn't change any parameter names to avoid introducing mistakes.

Value head with cross entropy training: https://nofile.io/f/w1s6m9fdQ65/CE-networks.zip
Value head with normal MSE training: https://nofile.io/f/dr0rKKUEqJq/MSE-networks.zip

Logs for CE: https://nofile.io/f/bM6TOQetKs3/leelalogs-CE.zip
Logs for MSE: https://nofile.io/f/9oVaOK4cAVa/leelalogs-MSE.zip

**Comment by gcp** [comment]
Thanks for running the tests. It seems it basically does not make a big difference in the end?

**Comment by tterava** [comment]
Using cross entropy looked very much like using MSE with higher learning rate. Initially it converged faster but after more steps it fell behind. I didn't see any crazy results where one would be significantly better.

It is possible that CE would give better results if the learning rate was tuned down, especially when a network is close to its limit. However I don't really know how to test this. I'm happy to run additional tests if you figure out a way to get some conclusive results. It's very hard to analyze the results when both learning rate and loss function cannot be directly compared. Maybe one way would be to tinker with the learning rate of CE and possible find a value where the CE-trained network performs better than MSE-trained network at any amount of training steps. I think that would conclusively show that CE is better. I can give that a try but it will probably take a long time to run all the tests.

**Comment by alreadydone** [comment]
https://github.com/lightvector/GoNN#cross-entropy-vs-l2-value-head-loss-aug-2018

**Comment by sethtroisi** [comment]
closing >6 month old issue with no clear owner.
